### ISSUE TITLE
Deeply nested dependencies are not copied

### ISSUE THREAD

**Issue by mpsijm** [zbigg/copy-production-deps#359]
To reproduce:
```
~ $ mkdir tmp
~ $ cd tmp
~/tmp $ npm i --save mongoose@7.6.4
~/tmp $ npm i --save-dev copy-production-deps
~/tmp $ npm ls @mongodb-js/saslprep
tmp@ /home/mpsijm/tmp
└─┬ mongoose@7.6.4
  └─┬ mongodb@5.9.0
    └── @mongodb-js/saslprep@1.1.1

~/tmp $ mkdir dist
~/tmp $ npx copy-production-deps
~/tmp $ cd dist
~/tmp/dist $ npm ls @mongodb-js/saslprep
/home/mpsijm/tmp/dist
└── (empty)

```
In this example, the nested dependency `@mongodb-js/saslprep` should be copied to `dist/node_modules` (as dependency of `mongodb`, but it's not. Note that `mongodb` itself _is_ copied, even though its dependencies are not:
```
~/tmp/dist $ npm ls mongodb
/home/mpsijm/tmp/dist
├── mongodb@5.9.0 extraneous
└─┬ mongoose@7.6.4 extraneous
  └── mongodb@5.9.0 deduped

```
This is probably because `mongodb` exists at the top-level `node_modules` folder, whereas `@mongodb-js/saslprep` does not.

System info:
- Arch Linux
- Node v18.18.0
- NPM v10.2.1

Note: this is a duplicate of https://github.com/arloliu/copy-node-modules/issues/42. I was looking for alternatives that had more recent releases than 5 years ago, but unfortunately, this package has the same bug :sweat_smile:

**Comment by mpsijm** [comment]
As a workaround, it is also possible to copy the entire repository, run `npm i --omit dev`, and then copy the resulting `node_modules` folder.

**Comment by zbigg** [comment]
Thanks for report. Unfortunately i wasn't able to look at it eariler.

Looks like saslprep is optionalDependency of mongodb:

```
// node_modules/mongodb/package.json
 "optionalDependencies": {
    "@mongodb-js/saslprep": "^1.1.0"
  },
 ```

Frankly, when i was writing this utility i wasn't aware of full semantics of `optionalDependecies` and thus one can say, that as for now `copy-production-deps` treats them as non-mandatory and doesn't copy them.

The workaround i've found is to add this
   "@mongodb-js/saslprep": "*"

to `dependencies` in `package.json` of your project.

**Comment by zbigg** [comment]
So, after reading how `optionalDependecies` do work, they are optional in sense `npm/yarn` always attempts to install them but doesn't fail if they are impatible/unavailable/broken (?).

So `copy-production-deps` should just add these to dep tree, but ignore if they aren't properly instantiated in `node_modules` by npm/yarn assuming they did best effort to install them.

**Comment by mpsijm** [comment]
> So `copy-production-deps` should just add these to dep tree, but ignore if they aren't properly instantiated in `node_modules` by npm/yarn assuming they did best effort to install them.

Which is interesting, because `@mongodb-js/saslprep` _is_ instantiated in `node_modules` in the reproduction example :thinking: Could the bug be caused by the fact that this package contains a slash, and is thus installed in the folder `node_modules/@mongodb-js/`?

**Comment by zbigg** [comment]
It's not slash and it's not interesting at all. `copy-production-deps` performs "own" analysis of what is needed and copies only these packages/sub-packages that it finds needed.
Note, in particular it doesn't blindly copy `node_modules/x/node_modules`, as it's possible that those dependencies in production can be resolved already from root parent `node_modules`, whereas in development not.

In this case we just completely ignore `optionalDependencies` because of reasons mentioned above.

**Comment by zbigg** [comment]
@mpsijm supposedly 🙏 fixed in just released `0.2.2`
